### PR TITLE
Add rpc.api._barrier()

### LINKED ISSUE #40166: Expose rpc._wait_all_workers as rpc.barrier
RPC uses `_wait_all_workers` as a barrier to do graceful shutdown. This might be a useful feature to applications as well, i.e., we can have `rpc.barrier()` for RPC and `dist.barrier()` for c10d. 

Note that this is different from #40107:

* The `rpc.barrier()` only makes sure that all workers reach the same point, while there could still be on-going concurrent execution in other RPC threads to satisfy futures created by the main thread. 
* The context manager collects and waits for all futures created in scope, but does not communicate with other workers.

https://github.com/pytorch/pytorch/blob/3fb1e73a4e928a0452c3b73ca6815f67127424e9/torch/distributed/rpc/api.py#L137-L198

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @jjlilley @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @xush6528

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53423 Add rpc.api._barrier()**

closes #40166

This change exposes a new internal API, rpc.api._barrier() which blocks the processes of workers running RPC until the specified group of workers completes this function.

Example:
```python
import os
import torch.multiprocessing as mp
import torch.distributed.rpc as rpc
os.environ["MASTER_ADDR"] = "localhost"
os.environ["MASTER_PORT"] = "5678"

world_size = 4
odd_num_workers = [f"worker{i}" for i in range(world_size) if i % 2]
even_num_workers = [f"worker{i}" for i in range(world_size) if not i % 2]

def worker(i):
    print(i)
    rpc.init_rpc(f"worker{i}", rank=i, world_size=world_size)
    if i % 2:
        print(f"start barrier on {i} for {odd_num_workers}")
        rpc.api._barrier(odd_num_workers)
    else:
        print(f"start barrier on {i} for {even_num_workers}")
        rpc.api._barrier(even_num_workers)
    rpc.shutdown()
    print(f"shutdown{i}")

if __name__ == '__main__':
    with mp.Pool(processes=world_size) as pool:
        pool.map(worker, range(world_size))
```

Output:

```
0
1
2
3
start barrier on 0 for ['worker0', 'worker2']
start barrier on 1 for ['worker1', 'worker3']
start barrier on 2 for ['worker0', 'worker2']
start barrier on 3 for ['worker1', 'worker3']
shutdown2
shutdown3
shutdown1
shutdown0
```

Differential Revision: [D27737145](https://our.internmc.facebook.com/intern/diff/D27737145)

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
This would be quite useful to have especially for RPC debugability, are there still plans to iterate on this? @H-Huang

**Comment by H-Huang** [comment]
Yes, I'll finish the changes needed for this

File: torch/distributed/rpc/api.py [hunk 1]
```diff
@@ -172,38 +173,47 @@ def _wait_all():
...
-        _all_gather_sequence_id += 1
 
     is_leader = leader_name == self_name
     if timeout == UNSET_RPC_TIMEOUT:
         timeout = get_rpc_timeout()
 
+    if worker_names != _ALL_WORKER_NAMES:
+        # create hash of all worker names in group
+        concat_names = "".join(sorted(worker_names)).encode('utf-8')
+        sequence_id = hashlib.sha256(concat_names).hexdigest()
```
Comment (H-Huang) [inline torch/distributed/rpc/api.py:201]: There is a race condition when multiple processes using the same sequence ID corresponding to the same set of workers, also need to make this a counter, set(workers) -> sequence_id

File: torch/distributed/rpc/__init__.py [hunk 2]
```diff
@@ -198,6 +198,25 @@ def init_rpc(
...
+    def barrier(worker_names=None):
+        r"""
+        Synchronizes local and remote RPC processes.
+
+        This will block until all local and remote RPC processes specified under worker_names
+        reach this method to wait for all outstanding work to complete.
+
+        Args:
+            worker_names (List[str], optional): The set of workers to synchronize. If ``None``, the
+            set will be all workers. Default is ``None``.
```
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:211]: we might need one more indent on the second line. Shall we build the docs to verify this can render correctly?

Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:211]: > If ``None``, the set will be all workers. Default is ``None``.

This will be a little tricky after we add dynamic membership. As that would require all processes always have a consistent view on the current gang membership, which can be costly and hard. To avoid that potential problem, shall we require `worker_names` to be not None for now?

Comment (mrshenli) [reply torch/distributed/rpc/__init__.py:211]: Two Questions: 

1. what if there are concurrent barriers, e.g., two current barriers between [work0, work1] and [worker2, worker3]. Will this still work?
2. If one of the worker specified a wrong list, what gonna happen?

File: torch/distributed/rpc/__init__.py [hunk 3]
```diff
@@ -198,6 +198,25 @@ def init_rpc(
...
 
+    @api._require_initialized
+    def barrier(worker_names=None):
+        r"""
+        Synchronizes local and remote RPC processes.
+
+        This will block until all local and remote RPC processes specified under worker_names
+        reach this method to wait for all outstanding work to complete.
+
+        Args:
```
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:209]: Shall we add an example section to this API?

**Review by mrshenli — COMMENTED** [review]
Looping in @kiukchung to make sure this does not conflicts with the control barrier idea. IIUC, rpc.barrier is irrelevant to the control barrier @kiukchung proposed and both can be useful. Just wanna double check. :)

File: torch/distributed/rpc/api.py [hunk 4]
```diff
@@ -172,38 +173,41 @@ def _wait_all():
...
-    global _all_gather_sequence_id
     with _all_gather_dict_lock:
-        sequence_id = _all_gather_sequence_id
-        _all_gather_sequence_id += 1
-
+        concat_names = "".join(sorted(worker_names))
+        sequence_num = _all_gather_sequence_id.get(concat_names, 0)
+        _all_gather_sequence_id[concat_names] = sequence_num + 1
+        sequence_id = hashlib.sha256((concat_names + str(sequence_num)).encode("utf-8")).hexdigest()
+
```
Comment (mrshenli) [inline torch/distributed/rpc/api.py:198]: lint error is real

**Review by mrshenli — COMMENTED** [review]
changes LGTM in general. The main concern I had is that, once we expose this as a public API, we will need to have better coverage for corner cases and error handlings. Could you please add some tests to make sure that when one of the workers specifies a wrong list of worker names, our error message is reasonable? Concurrent barrier might be harder to support. If we don't plan to do it in the first version, let's make it clear in the doc. 

@rohan-varma do you know who might be interested in trying this API?

File: torch/distributed/rpc/__init__.py (hunk 2 above)
Comment (H-Huang) [reply torch/distributed/rpc/__init__.py:211]: Great points. I will change to the worker_names to be a required argument. For the questions:

1. Yes, this works.
2. If the list has workers that doesn't exist, then it would hang and eventually timeout.

**Comment by kiukchung** [comment]
> Looping in @kiukchung to make sure this does not conflicts with the control barrier idea. IIUC, rpc.barrier is irrelevant to the control barrier @kiukchung proposed and both can be useful. Just wanna double check. :)

Thanks for looping me in @mrshenli! Shouldn't conflict, but semantically `torch.runtime.barrrier()`would do exactly the same thing with the only advantage being that it does not depend on RPC (nor process groups) and rather depends on the the elastic agent (separate process), making it more robust.

File: torch/distributed/rpc/__init__.py (hunk 2 above)
Comment (mrshenli) [reply torch/distributed/rpc/__init__.py]: > If the list has workers that doesn't exist, then it would hang and eventually timeout.

I was also concerned about the following case:

worker0: `worker_names=["worker0", "worker1", "worker2"]`
worker1: `worker_names=["worker0", "worker1", "worker2"]`
worker2: `worker_names=["worker1", "worker2"]`

Do we know what gonna happen in this case? I guess it will also timeout as `worker0` serves as the leader for the first two and `worker1` serves as the leader for `worker2`.

Comment (H-Huang) [reply torch/distributed/rpc/__init__.py]: Yes, this situation would also time out, which I think should be expected behavior of the API? For the implementation right now, if one worker calls `barrier(<list_of_workers>)` it expects all the workers in <ilst_of_workers> to also call `barrier(<list_of_workers>)` otherwise it will hang.

**Comment by rohan-varma** [comment]
> @rohan-varma do you know who might be interested in trying this API?

Sorry, missed this comment. I'm not aware of any immediate users, but I think it would be pretty invaluable for folks debugging their applications when looking at possible correctness issues/race conditions.

File: torch/distributed/rpc/api.py [hunk 5]
```diff
@@ -116,19 +114,21 @@ def _init_rpc_states(agent):
...
+def _gather_to_leader(sequence_id, worker_name, obj, worker_names=None):
     with _all_gather_dict_lock:
-        assert (
-            worker_name in _ALL_WORKER_NAMES
-        ), "{worker_name} is not expected by leader.".format(worker_name=worker_name)
+        if not worker_names:
+            worker_names = _ALL_WORKER_NAMES
+            assert (
+                worker_name in worker_names
+            ), "{worker_name} is not expected by leader.".format(worker_name=worker_name)
```
Comment (rohan-varma) [inline torch/distributed/rpc/api.py:123]: nit: f-string?

File: torch/distributed/rpc/api.py [hunk 6]
```diff
@@ -238,9 +241,31 @@ def _all_gather(obj, timeout=UNSET_RPC_TIMEOUT):
...
             )
 
+    # Clean up for the states using the sequence_id
+    with _all_gather_dict_lock:
+        states = _all_gather_sequence_id_to_states.pop(sequence_id)
     return states.gathered_objects
 
 
+@_require_initialized
+def _barrier(worker_names):
```
Comment (rohan-varma) [inline torch/distributed/rpc/api.py:251]: Should we make it a public API?

File: torch/distributed/rpc/api.py [hunk 7]
```diff
@@ -238,9 +241,31 @@ def _all_gather(obj, timeout=UNSET_RPC_TIMEOUT):
...
+
+    Args:
+        worker_names (List[str]): The set of workers to synchronize.
+
+    """
+    try:
+        _all_gather(None, set(worker_names), timeout=DEFAULT_SHUTDOWN_TIMEOUT)
+    except RuntimeError as ex:
+        logger.error(
+            f"Failed to respond to 'Shutdown Proceed' in time, got error {ex}"
```
Comment (rohan-varma) [inline torch/distributed/rpc/api.py:266]: If this is being used for non-shutdown case, this error might be a bit confusing (says "shutdown proceed")?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 8]
```diff
@@ -1055,6 +1055,35 @@ def test_all_gather_timeout(self):
             with self.assertRaisesRegex(RuntimeError, expected_error):
                 rpc.api._all_gather(SlowPickleClass(0.5))
 
+    @dist_init
+    def test_rpc_barrier_all(self):
+        info = rpc.get_worker_info()
+        all_worker_info = rpc._get_current_rpc_agent().get_worker_infos()
+        names = [worker.name for worker in all_worker_info]
+        rpc.api._barrier(names)
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:1063]: Ideally a test should assert something to test the functionality, though admittedly it is a bit hard for barrier. Could we do something like, before the barrier, assert some global count is 0, then send RPC to each node to increment the global count, and then call barrier, then assert the count == world_size? This will ensure barrier properly waits for all outstanding work. Open to better ideas though.

File: torch/distributed/rpc/api.py [hunk 9]
```diff
@@ -172,38 +172,41 @@ def _wait_all():
...
+    leader_name = sorted(worker_names)[0]
 
     self_name = _get_current_rpc_agent().get_worker_info().name
 
-    global _all_gather_sequence_id
     with _all_gather_dict_lock:
-        sequence_id = _all_gather_sequence_id
-        _all_gather_sequence_id += 1
+        concat_names = "".join(sorted(worker_names))
+        sequence_num = _all_gather_sequence_id.get(concat_names, 0)
```
Comment (rohan-varma) [inline torch/distributed/rpc/api.py:194]: I see that we properly clean up this key for `_all_gather_sequence_id` below when exiting the barrier so that future calls with the same set of worker_names will start at 0 as expected.

However, is it possible to still run into races here? If there are concurrent calls to `barrier()` (i.e. multiple threads) with the same set of workers:

Thread 1 creates the key in `_all_gather_sequence_id`
Thread 1 releases lock 
Thread 2 calls with the same set of arguments, so it sees an already incremented sequence_num, which could cause the tracking of states to be incorrect.

File: torch/distributed/rpc/api.py (hunk 6 above)
Comment (H-Huang) [reply torch/distributed/rpc/api.py:251]: I had discussed this with Shen a few weeks ago as the PR was originally for public API. But since we are unsure of the people who will use it and also how the API may change when introducing Elastic RPC, decided to keep it internal for now.

File: torch/distributed/rpc/api.py (hunk 7 above)
Comment (H-Huang) [reply torch/distributed/rpc/api.py:266]: Good catch, I think i had copy-pasted from shutdown() 🙂, will change it.

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 8 above)
Comment (H-Huang) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:1063]: That's a good point. I'll implement something along the count idea. I'll also add something to test the negative case where barrier does actually time out.

File: torch/distributed/rpc/api.py (hunk 9 above)
Comment (H-Huang) [reply torch/distributed/rpc/api.py:192]: The cleanup done at the end of barrier is for `_all_gather_sequence_id_to_states`. This var keeps tracks of the barrier state for a single sequence id. This cleanup is needed because there is an existing memory leak (small one) where each call to `_all_gather` will leave behind an `AllGatherStates` class in the global dictionary.

The `_all_gather_sequence_id` is an ID representing a particular barrier operation. It is strictly increasing so that concurrent calls will still work. I'll rewrite your example above to demonstrate what is happening:

- worker1, Thread 1 performs barrier on ["worker1", "worker2"], retrieve `_all_gather_sequence_id` (value 0), increment `_all_gather_sequence_id`
- worker1, Thread 1 release lock
- worker1, Thread 2 performs barrier on ["worker1", "worker2"], retrieve `_all_gather_sequence_id` (value 1), increment `_all_gather_sequence_id`
- Thread1 and Thread2 block waiting on the barrier from worker2
- When worker2 calls barrier ["worker1", "worker2"] the first time, it will allow worker1, Thread1 to make progress. The second call to barrier will allow worker1, Thread2 to make progress.

Comment (rohan-varma) [reply torch/distributed/rpc/api.py:192]: Sounds good. Could we add a unittest to verify this multithreaded behavior?

File: torch/distributed/rpc/api.py [hunk 10]
```diff
@@ -238,9 +239,31 @@ def _all_gather(obj, timeout=UNSET_RPC_TIMEOUT):
...
+@_require_initialized
+def _barrier(worker_names):
+    r"""
+    Synchronizes local and remote RPC processes.
+
+    This will block until all local and remote RPC processes specified under worker_names
+    reach this method to wait for all outstanding work to complete.
+
+    Args:
+        worker_names (List[str]): The set of workers to synchronize.
```
Comment (rohan-varma) [inline torch/distributed/rpc/api.py:257]: Can be added in a follow up diff, but should we also accept WorkerInfo or int for WorkerId here? The rest of the RPC APIs seem to accept either so might be good to add that.

**Review by rohan-varma — APPROVED** [review]
Looks great! Thanks for adding the multithreading tests as we discussed.

File: torch/distributed/rpc/api.py (hunk 10 above)
Comment (H-Huang) [reply torch/distributed/rpc/api.py:257]: Sounds good, will do.

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27737145).
